### PR TITLE
Add server package

### PR DESCRIPTION
**Pull request by daniel-fanjul-alcuten**

### CODE REVIEW

File: listen/listen.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	Listener net.Listener
+	Timeout  time.Duration
+	Serve    func(net.Conn)
+}
+
+type deadlineSettable interface {
+	SetDeadline(t time.Time) error
+}
+
+// Listen calls l.Listener.Accept() in a loop until the context l.Ctx is done
```
Comment (manuelbcd) [inline listen/listen.go:21]: May I suggest to comment with /* ... */ for multi line coments ?

File: listen/listen.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// or there is an error. time.Now().Add(l.Timeout) is passed to
+// l.Listener.SetDeadline() before each call to Accept(), and errors caused by
+// this deadline are ignored. Every accepted connection is passed to the
+// callback l.Serve() in a new goroutine that closes it afterwards.
+func (l *Listener) Listen() error {
+	ds := l.Listener.(deadlineSettable)
+	for {
+		select {
```
Comment (manuelbcd) [inline listen/listen.go:29]: wrong indent?

File: listen/listen.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+// Listen calls l.Listener.Accept() in a loop until the context l.Ctx is done
+// or there is an error. time.Now().Add(l.Timeout) is passed to
+// l.Listener.SetDeadline() before each call to Accept(), and errors caused by
+// this deadline are ignored. Every accepted connection is passed to the
+// callback l.Serve() in a new goroutine that closes it afterwards.
+func (l *Listener) Listen() error {
+	ds := l.Listener.(deadlineSettable)
+	for {
```
Comment (manuelbcd) [inline listen/listen.go:28]: question... Is a good practice to put a blind 'for'?

File: server/server.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package server
+
+import (
+	"context"
+	"net"
+	"time"
+
```
Comment (manuelbcd) [inline server/server.go:7]: Is this empty line accepted?

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	if err != nil {
+		return err
+	}
+	defer l.Close()
+	ll := &listen.Listener{
+		Ctx:      s.Ctx,
+		Listener: l,
+		Timeout:  s.Timeout,
+		Serve:    s.Serve,
+	}
```
Comment (manuelbcd) [inline server/server.go:34]: I would suggest to add an empty line before main return... but I'm not sure if this is matching the standard

File: listen/listen.go (hunk 1 above)
Comment (daniel-fanjul-alcuten) [reply listen/listen.go:21]: No.
https://golang.org/doc/effective_go.html?#commentary

"Line comments are the norm; block comments appear mostly as package comments"

File: listen/listen.go (hunk 2 above)
Comment (daniel-fanjul-alcuten) [reply listen/listen.go:29]: No. I always use `gofmt`.

File: listen/listen.go (hunk 3 above)
Comment (daniel-fanjul-alcuten) [reply listen/listen.go:28]: It is neither good nor bad.
https://golang.org/ref/spec#For_statements

"If the condition is absent, it is equivalent to the boolean value true."

File: server/server.go (hunk 4 above)
Comment (daniel-fanjul-alcuten) [reply server/server.go:7]: Yes. I used `goimports` before `gofmt`.
https://github.com/golang/go/issues/8963

File: server/server.go (hunk 5 above)
Comment (daniel-fanjul-alcuten) [reply server/server.go:34]: There is no standard AFAIK, but I do not see why that would be better. This initializes a struct and uses it in the next line. A blank line there would be like a comma between subject and verb in a sentence.

**Pull request merged**